### PR TITLE
Adding "spectral clipped" weight decay to be used together with matrix-based preconditioning optimizers

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
Based on @[leloykun](https://github.com/leloykun)'s spectral clipping [algorithm](https://leloykun.github.io/ponder/spectral-clipping/)

### CODE REVIEW

File: emerging_optimizers/orthogonalized_optimizers/muon_utils.py [hunk 1]
```diff
@@ -36,14 +36,14 @@
...
     "polar_express": [
         # Polar Express iteration from: https://arxiv.org/abs/2505.16932
-        (7.2086, -15.5131, 9.0178),
-        (3.9623, -2.5813, 0.4542),
-        (3.9466, -2.5765, 0.4544),
-        (3.8991, -2.5671, 0.4566),
-        (3.7186, -2.5308, 0.4653),
-        (3.1390, -2.3073, 0.4733),
-        (2.1715, -1.5246, 0.3885),
-        (1.8648, -1.2224, 0.3577),
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_utils.py:46]: Make a separate PR for coefficient only change next time, in which case review can be very light.

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        The spectral clipped tensor.
+    """
+    if needs_transpose := X.shape[0] > X.shape[1]:
+        X = X.T
+    OX = newton_schulz(X, steps=8, coefficient_type="polar_express")
+    result = (1 / 2) * (
+        (sigma_min + sigma_max) * OX
+        + newton_schulz(sigma_min * torch.eye(X.shape[0]) - OX @ X.T, steps=8, coefficient_type="polar_express")
+        @ (sigma_min * OX - X)
+        - newton_schulz(sigma_max * torch.eye(X.shape[0]) - OX @ X.T, steps=8, coefficient_type="polar_express")
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: Line too long and DRY, very hard to read.

Make it a loop. use intermediate variable to make line shorter.

AI give me this. 
```python
    ns_kwargs = dict(steps=8, coefficient_type="polar_express")
    OX = newton_schulz(X, **ns_kwargs)
    result = (sigma_min + sigma_max) * OX
    I = torch.eye(X.shape[0], device=X.device, dtype=X.dtype)
    for s, sign in zip([sigma_min, sigma_max], [1, -1]):
        A = s * I - OX @ X.T
        B = s * OX - X
        result += sign * newton_schulz(A, **ns_kwargs) @ B
    result = 0.5 * result
```
you can try to improve it further. use add with alpha can also improve performance

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        @ (sigma_max * OX - X)
+    )
+    if needs_transpose:
+        result = result.T
+    return result
+
+
+def spectral_hardcap(X: torch.Tensor, beta: float = 1.0) -> torch.Tensor:
+    r"""Spectral hardcap function clips singular values from above to be less than beta.
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py:56]: from "larger" to be "smaller" than?

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    Returns:
+        The spectral hardcapped tensor.
+
+    """
+    if needs_transpose := X.shape[0] > X.shape[1]:
+        X = X.T
+    OX = newton_schulz(X, steps=8, coefficient_type="polar_express")
+    aX = beta * OX - X
+    result = (1 / 2) * (beta * OX + X - aX @ newton_schulz(aX, steps=8, coefficient_type="polar_express").T @ OX)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: Same as before, very hard to read line. Break it down.

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    Args:
+        X: The input tensor.
+        beta: The upper bound on the singular values.
+        c: The coefficient parameter.
+
+    Returns:
+        The spectral clipped weight decay tensor.
+    """
+    return (1 - c) * X + c * spectral_hardcap(X, beta)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: Use torch.add to save some memory traffic.

File: tests/test_spectral_clip_utils.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from absl import logging
+from absl.testing import absltest, parameterized
+
+from emerging_optimizers.orthogonalized_optimizers.spectral_clipping_utils import (
```
Comment (skyw) [inline tests/test_spectral_clip_utils.py]: suggestion: Follow Google style guide, import module only, not function.

File: tests/test_spectral_clip_utils.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        original_max_sv = original_singular_values.max().item()
+
+        clipped_x = spectral_clip(x, sigma_min=sigma_min, sigma_max=sigma_max)
+
+        _, singular_values, _ = torch.linalg.svd(clipped_x, full_matrices=False)
+
+        min_sv = singular_values.min().item()
+        max_sv = singular_values.max().item()
+
+        logging.info(f"Original matrix shape: {x.shape}")
```
Comment (skyw) [inline tests/test_spectral_clip_utils.py]: This is too verbose, can leave them as debug.

File: tests/test_spectral_clip_utils.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        logging.info(f"Original singular values range: [{original_min_sv:.6f}, {original_max_sv:.6f}]")
+        logging.info(f"Clipped singular values range: [{min_sv:.6f}, {max_sv:.6f}]")
+        logging.info(f"Target range: [{sigma_min:.6f}, {sigma_max:.6f}]")
+        logging.info(f"Shape preservation: input {x.shape} -> output {clipped_x.shape}")
+
+        tolerance_upper = 1e-1
```
Comment (skyw) [inline tests/test_spectral_clip_utils.py:63]: Comment how tolerance is chosen and why upper and lower are so different.

File: tests/test_spectral_clip_utils.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        self.assertEqual(clipped_x.shape, x.shape)
+
+    @parameterized.product(
+        shape_pairs=[(256, 128), (128, 256), (512, 512), (100, 200)],
+        beta_values=[0.5, 1.0, 0.8, 2.0],
+    )
+    def test_spectral_hardcap(self, shape_pairs, beta_values):
+        """Test that spectral hardcap properly clips singular values from above to be less than beta."""
+        dim1, dim2 = shape_pairs
+        beta = beta_values
```
Comment (skyw) [inline tests/test_spectral_clip_utils.py]: Meaningless reassign. the name `beta_values` and `shape_pairs` shouldn't be plural. They should be named w.r.t to test case, not in parameterized where they are lists.

File: tests/test_spectral_clip_utils.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        logging.info(f"Polar factor Frobenius norm difference: {relative_polar_frobenius_diff:.6f}")
+
+        self.assertLessEqual(
+            relative_polar_frobenius_diff,
+            polar_tolerance,
+        )
+
+
+if __name__ == "__main__":
+    absltest.main()
```
Comment (skyw) [inline tests/test_spectral_clip_utils.py:126]: Also add the test file to CI scipt.

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py (hunk 3 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py:56]: yes

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+    Returns:
+        The spectral clipped tensor.
+    """
+    if needs_transpose := X.shape[0] > X.shape[1]:
+        X = X.T
+    OX = newton_schulz(X, steps=8, coefficient_type="polar_express")
+    result = (sigma_min + sigma_max) * OX
+    identity_matrix = torch.eye(X.shape[0], device=X.device, dtype=X.dtype)
+    # Pre-allocate tensors for memory efficiency
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: Actually it doesn't, and add visual noise.
`torch.empty(A)` is called internally for `A = torch.add(s * identity_matrix, OX @ X.T, alpha=-1)`

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    A = torch.empty_like(identity_matrix)
+    B = torch.empty_like(X)
+    for s, sign in zip([sigma_min, sigma_max], [1, -1]):
+        torch.add(s * identity_matrix, OX @ X.T, alpha=-1, out=A)
+        torch.add(s * OX, X, alpha=-1, out=B)
+        torch.add(result, sign * newton_schulz(A, steps=8, coefficient_type="polar_express") @ B, out=result)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: either return to result or use result.add_ if really want to be inplace. Use out=result to indicate inplace is very hard to read and can be easily missed.

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+    Returns:
+        The spectral hardcapped tensor.
+
+    """
+    if needs_transpose := X.shape[0] > X.shape[1]:
+        X = X.T
+    OX = newton_schulz(X, steps=8, coefficient_type="polar_express")
+    aX = torch.add(beta * OX, X, alpha=-1)
+    result = torch.empty_like(X)
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: same comment as in spectral_clip

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py (hunk 12 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: Returned to result instead

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py (hunk 13 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: returned to result and removed torch.empty

File: emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py (hunk 11 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/orthogonalized_optimizers/spectral_clipping_utils.py]: removed` torch.empty `
